py-cryptography: Update to 3.4.7

essandess opened the pull request:
#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'printf "%s\n" "macOS `sw_vers -productVersion` `sw_vers -buildVersion` `uname -m`" "`xcodebuild -version|awk '\''NR==1{x=$0}END{print x" "$NF}'\''`"'|tee /dev/tty|pbcopy
-->
macOS x.y
Xcode x.y

###### Verification <!-- (delete not applicable items) -->
Have you

- [ ] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [ ] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [ ] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [ ] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

stromnov commented:
cryptography 3.x is backward incompatible with macOS 10.9 and below. So it's not possible to simply bump the library version without affecting old macOSes.

https://lists.macports.org/pipermail/macports-dev/2020-October/042505.html

essandess commented:
Let’s get on the major update for the great majority of deployments and backfill the old edge cases later. Perhaps create a port `py-cryptography2`.

stromnov commented:
As far as I see, there will be too many edge cases. Every port, which depends on py-cryptography, will be an edge case, requiring substitution py-cryptography with py-cryptography2 (or -legacy) if OS version <= 10.9.

essandess commented:
This port is out-of-date with several known insecurities: https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst

May we please update ASAP?

If anyone wants to install insecure, out-of-date versions they may choose to do so, but MacPorts shouldn’t be distributing cryptography code with unpatched CVEs.

kencu commented:
builds are failing due to:
```
2021-04-28T22:31:15.2833650Z ModuleNotFoundError: No module named 'setuptools_rust'
2021-04-28T22:31:15.2967450Z Command failed:  cd "/opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_python_py-cryptography/py38-cryptography/work/cryptography-3.4.7" && /opt/local/Library/Frameworks/Python.framework/Versions/3.8/bin/python3.8 setup.py --no-user-cfg build -j3 
```

First I've seen of "setuptools_rust" -- oh what fun, if python ports now need rust to build :>

stromnov commented:
> You will also need to have Rust installed and available, which can be obtained from Homebrew, MacPorts, or directly from the Rust website.

https://cryptography.io/en/latest/installation/#building-cryptography-on-macos

herbygillot commented:
> First I've seen of "setuptools_rust" -- oh what fun, if python ports now need rust to build :>

Yep, this has caused plenty of Internet entertainment. You can probably start with  https://github.com/pyca/cryptography/issues/5771

essandess commented:
Looks like we should also remove Python 35 (and 36 too), but there are dependencies to chase down.

https://github.com/pyca/cryptography/blob/33bc4fd2394b0168508cff1e8a0c6645528d47de/setup.py#L90

I suggest we leave these for another PR and expunge the CVEs.

```bash
port echo rdepends:py35-cryptography
py35-ansible                    
py35-ansible-base               
py35-comet                      
py35-fido2                      
py35-libcloud                   
py35-oauthlib                   
py35-openssl                    
py35-paramiko                   
py35-requests-oauthlib          
py35-service_identity           
py35-trustme                    
py35-tweepy                     
py35-twilio
```

reneeotten commented:
this newer version is Python 3.6+ so it will need to take care of PY27 ports as well. There needs to be a plan in place for how to either remove all the (recursive) dependencies of these ports or pin them to an earlier version. I think that should be part of this PR and not postponed to a later point; merging this will likely break things and the required clean-ups might not happen anytime soon if not done now. Additionally, there is the point of breaking for 10.9 and below that @stromnov mentions - what's the opinion on that?

herbygillot commented:
Does `setuptools-rust` still need to be in `depends_lib-append`, or can it live in `depends_build-append`?

essandess commented:
> Does `setuptools-rust` still need to be in `depends_lib-append`, or can it live in `depends_build-append`?

I put it in `depends_build-append`, added `cargo_fetch` stuff, and added tests. The tests all run in Python 39.

reneeotten commented on python/py-cryptography/Portfile:
```diff
@@ -32,18 +33,93 @@ platform macosx {
+                    port:py${python.version}-asn1crypto \
+                    port:py${python.version}-cffi \
+                    port:py${python.version}-pip \
+                    port:py${python.version}-setuptools \
```
why is `setuptools` not only a build dependency; and were do `py-six` and `py-pip` come from, I don't see them listedupstream, or...?

essandess replied:
I put `setuptools` in `depends_build`. I put `pip` in there because `port -vst install` failed without it and complained explicitly about not finding `pip`.

reneeotten replied:
sure I was suggesting to put `py-setuptools` in `depends_build` instead of `depends_lib` where you had it initially since it shouldn't be required at runtime. I still don't see where `pip` comes in, that shouldn't be needed I would think; please show the output of the build failure.

essandess replied:
I'm honestly not sure about that—seems like it was necessary with trace mode. I'll delete it and hope for the best.

essandess replied:
These trace mode and CI failures are why I believe that a pip build dependency is required:
```
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/setuptools/installer.py", line 75, in fetch_build_egg
    subprocess.check_call(cmd)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/opt/local/Library/Frameworks/Python.framework/Versions/3.9/bin/python3.9', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/opt/local/var/macports/build/_Users_runner_work_1_s_ports_python_py-cryptography/py39-cryptography/work/.tmp/tmpeb_zl60_', '--quiet', '--index-url', "''", '--find-links', "''", 'toml>=0.9.0']' returned non-zero exit status 1.
```

reneeotten reviewed (changes requested):
before a possible merge or this PR, this will need a solution for unsupported Python versions and possibly older systems and, importantly, agreement from the current maintainer.

reneeotten commented on python/py-cryptography/Portfile:
```diff
@@ -32,18 +33,93 @@ platform macosx {
+
+    depends_lib-append \
+                    path:lib/libssl.dylib:openssl \
+                    port:py${python.version}-asn1crypto \
```
also since version 2.8 it seems that `asn1crypto` is no longer a [dependency](https://github.com/pyca/cryptography/blob/b2f9a0e1a546c3b10a6e96cd5e8423a7b04970c7/CHANGELOG.rst#28---2019-10-16).

essandess replied:
Thanks fixed.

reneeotten commented on python/py-cryptography/Portfile:
```diff
@@ -32,18 +33,93 @@ platform macosx {
+if {${name} ne ${subport}
+    && [string match "py*-${github.project}" ${subport}]} {
+    depends_build-append \
+                    port:py${python.version}-setuptools-rust \
```
this only has subports for py37/38/39

essandess replied:
That’s what upstream supports, ignoring EOL 36:

https://github.com/PyO3/setuptools-rust/blob/61b0603ba8cecaff34331904de8f2a0d035152ca/setup.cfg#L31

reneeotten replied:
sure, but that means that you will need to remove the PY27/PY35 subports from `py-cryptography` (which doesn't support lower versions either) and add PY36 to `py-setuptools-rust`

reneeotten replied:
and Python 3.6 is not EOL [yet](https://devguide.python.org/#status-of-python-branches)

essandess replied:
Right, not dead yet. I kept it in the latest push.

reneeotten commented on python/py-cryptography/Portfile:
```diff
@@ -1,11 +1,13 @@
+PortGroup           cargo_fetch 1.0
+PortGroup           github 1.0
 PortGroup           python 1.0
 PortGroup           compiler_blacklist_versions 1.0
```
are you actually using this PG?

essandess replied:
“`cargo.crates`” avoids rust fetches after the fetch phase.

reneeotten replied:
sorry for not being completely clear, I was referring to the `compiler_blacklist_versions` PG

reneeotten replied:
I still don't see where the `compiler_blacklist_versions` PG is used in that subport either; I might be missing it, so please point me to it

essandess replied:
Removed thanks.

essandess commented:
> before a possible merge or this PR, this will need a solution for unsupported Python versions and possibly older systems and, importantly, agreement from the current maintainer.

Agreed that EOL Python dependencies should be removed. The major affected port looks to be mailman’s, but this port is out-of-date and must be updated to `mailman3`. I’ll defer to other whether to continue support for `mailman2` and how to accomplish that.

Older systems are already explicitly not supported by upstream—this happened four years ago and isn’t up to us: https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst#19---2017-05-29

The current version that MacPorts distributes has several known CVEs. We must update to the latest version as soon as possible.

kencu commented:
If this port is merged without any backup in place, and all systems that can't support rust are dropped, just what kind of rdeps wreckage is expected here?

One of MacPorts unique strengths is it is still useful to all those people out there who have older systems they are not ready to bin yet. It would be a shame to cut them all off because of what amounts to one guy's rampage about C/C++ being evil languages that need to be dropped in favour of rust.

So taking a bit of time to come up with a coherent plan is a good idea. There is nothing urgent about this.

kencu commented:
IMHO, of course. I clearly don't speak for MacPorts, nor have any particular sway in the mix.

essandess commented:
> If this port is merged without any backup in place

I just added legacy support. Old dependencies just need to point at `py*-cryptography2`.

kencu commented:
> Not serving up CVEs takes precedence many times over the tradeoff of supporting old systems that are explicitly not supported upstream. This isn’t MacPorts pulling the plug, upstream pulled it years ago. It’s out of our hands.

No debate about that.

But leaving behind the last non-rust version for those of us who don't work in major multinational bank or at the NSA, and who just want our kids to be able to play some games on the old computer or use it to read some books, is not perhaps a terribly big ask?

reneeotten commented:
> > If this port is merged without any backup in place
> 
> I just added legacy support. Old dependencies just need to point at `py*-cryptography2`.

please make a separate port for `py-cryptography2`; just putting everything into one file doesn't help readability/maintainability.

I agree with @kencu that having a solid plan first before merging this makes sense. This port has been outdated for years without anyone being bothered about it - so a few more days or a week shouldn't be an issue. 

So what's the plan forward now: are you suggesting to change all dependents of this port to `py-cryptography2` for systems that cannot build `rust`? If so, please do that in this PR.

essandess commented:
> So what's the plan forward now: are you suggesting to change all dependents of this port to `py-cryptography2` for systems that cannot build `rust`? If so, please do that in this PR.

Please see #10873.

reneeotten commented:
> > So what's the plan forward now: are you suggesting to change all dependents of this port to `py-cryptography2` for systems that cannot build `rust`? If so, please do that in this PR.
> 
> Please see #10873.

Thanks @essandess - this will take care of Python 2.7 ports to use the latest supported version of `py-cryptography`. It doesn't deal yet with the fact that older systems that cannot install `rust` will have no way of installing any of the Python 3.x subports anymore - correct? 

I think that's the other remaining point that also @kencu and @stromnov  pointed out. To be able to build these Python ports on older systems, I think one would need to do the dependency replacement when `if {{${python.version} eq 27} or {${os.platform} eq "darwin" && ${os.major} < XY}}`, where `XY` is the latest version where `rust` builds.

essandess commented:
> > > So what's the plan forward now: are you suggesting to change all dependents of this port to `py-cryptography2` for systems that cannot build `rust`? If so, please do that in this PR.
> > 
> > 
> > Please see #10873.
> 
> Thanks @essandess - this will take care of Python 2.7 ports to use the latest supported version of `py-cryptography`. It doesn't deal yet with the fact that older systems that cannot install `rust` will have no way of installing any of the Python 3.x subports anymore - correct?
> 
> I think that's the other remaining point that also @kencu and @stromnov pointed out. To be able to build these Python ports on older systems, I think one would need to do the dependency replacement when `if {{${python.version} eq 27} or {${os.platform} eq "darwin" && ${os.major} < XY}}`, where `XY` is the latest version where `rust` builds.

My request is that further legacy details be addressed later because it involves a lot of MacPorts system details with which I am unfamiliar, takes a lot of time to rebase this many commits, and further delays elliminating the version 2 CVE from the MacPorts distribution.

essandess commented:
Re: `if {{${python.version} eq 27} or {${os.platform} eq "darwin" && ${os.major} < XY}}`

Upstream [says](https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst#30---2020-07-20) <blockquote>**BACKWARDS INCOMPATIBLE:** Dropped support for macOS 10.9, macOS users must upgrade to 10.10 or newer.</blockquote>

What's the `${os.major}` for macOS 10? If you all agree on the exact logic, I volunteer to rebase the commits in #10873 to use this logic.

reneeotten commented:
> Update: Please see all the commits in #10873. This is essentially represented in all the relevant dependencies:
> 
> ```
>     if {${python.version} eq 27} {
>         depends_lib-replace \
>                     port:py${python.version}-cryptography \
>                     port:py${python.version}-cryptography2
>     } 
> ```

yes, that will take care of using the previous version for PY27; but what happens if I would want to install the PY39 subport of one of these ports on a system that doesn't support `rust`?

essandess commented:
> Re: `if {{${python.version} eq 27} or {${os.platform} eq "darwin" && ${os.major} < XY}}`
> 
> Upstream [says](https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst#30---2020-07-20)
> 
> > **BACKWARDS INCOMPATIBLE:** Dropped support for macOS 10.9, macOS users must upgrade to 10.10 or newer.
> 
> What's the `${os.major}` for macOS 10? If you all agree on the exact logic, I volunteer to rebase the commits in #10873 to use this logic.

From [here](https://trac.macports.org/wiki/PortfileRecipes) it's `14`. I will use this logic in #10873 unless someone requests an alternate in the next 1–2 days:
```
    if {${python.version} eq 27 || {${os.platform} eq "darwin" && ${os.major} < 14}} {
        depends_lib-replace \
                    port:py${python.version}-cryptography \
                    port:py${python.version}-cryptography2
    } 
```

reneeotten commented:
@essandes - great, that would cover everything I think. From the [buildbots](https://ports.macports.org/port/rust/summary) it looks like `rust` did build on 10.9 so perhaps the latest version of that port can still be used there? Of course, I am not sure if there is something specifically in the `py-cryptography` port that would require 10.10+.

essandess commented:
The latest [force-push](https://github.com/macports/macports-ports/compare/4c5a2dc81bd8e7d944a663db03ead00ab3502527..4a111a52bcfed9211c5a11b25c7db175d0ff2504) implements the suggestions above and in #10873.

essandess commented:
@reneeotten These CI failures are why I believe that a pip build dependency is required:
```

Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/setuptools/installer.py", line 75, in fetch_build_egg
    subprocess.check_call(cmd)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/opt/local/Library/Frameworks/Python.framework/Versions/3.9/bin/python3.9', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/opt/local/var/macports/build/_Users_runner_work_1_s_ports_python_py-cryptography/py39-cryptography/work/.tmp/tmpeb_zl60_', '--quiet', '--index-url', "''", '--find-links', "''", 'toml>=0.9.0']' returned non-zero exit status 1.
```

reneeotten commented:
> @reneeotten These CI failures are why I believe that a pip build dependency is required:
> 
> ```
> Traceback (most recent call last):
>   File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/setuptools/installer.py", line 75, in fetch_build_egg
>     subprocess.check_call(cmd)
>   File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 373, in check_call
>     raise CalledProcessError(retcode, cmd)
> subprocess.CalledProcessError: Command '['/opt/local/Library/Frameworks/Python.framework/Versions/3.9/bin/python3.9', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/opt/local/var/macports/build/_Users_runner_work_1_s_ports_python_py-cryptography/py39-cryptography/work/.tmp/tmpeb_zl60_', '--quiet', '--index-url', "''", '--find-links', "''", 'toml>=0.9.0']' returned non-zero exit status 1.
> ```

That message is telling you that it is missing `toml` and tries to install that with `pip`, which fails because `pip` isn't there; the solution is to add `py-toml` to the build dependencies instead. Upstream has not listed this explicitly as a dependency, however it's pulled in indirectly through `setuptools-rust`; building in tracemode or on the buildbots without specifying it in this port will make the builds fail. (There are a few other remaining points that I'll review later).

reneeotten commented on python/py-pytest-subtests/Portfile:
```diff
@@ -0,0 +1,44 @@
+license             MIT
+fetch.type          git
+
+python.versions     37 38 39
```
this also needs a PY36 subport since it's needed for `py36-cryptography`, see the `lint` error: " Error: Unknown dependency: py36-pytest-subtests"

essandess replied:
Thanks done.

reneeotten commented on python/py-pytest-subtests/Portfile:
```diff
@@ -0,0 +1,44 @@
+platforms           darwin
+supported_archs     noarch
+license             MIT
+fetch.type          git
```
please take this from PyPI instead to avoid the `fetch.type git`

essandess replied:
Okay, but pretty sure that will break things.

Would you please step in and survey the breakage and offer a fix? There will be some complaint from the build command that it's not a valid git repo.

I could never figure that one out, so I just used a git clone to make this issue go away.

essandess replied:
Done. I predict a CI failure. Would you please take a look when it comes?

reneeotten replied:
you only removed `fetch.type git`, that will likely not work indeed. I suggested to take the file from PyPI, which *should* work correctly, but for that to happen you'll need to remove the `github` PG and associated stuff as well.

essandess replied:
I tried switching to PyPI before but it throws the same error. I'll amend to demonstrate…

essandess replied:
To the best of my knowledge, this is the last remaining major issue. I'm expecting a CI build failure using PyPI now…

essandess replied:
Looks like I was wrong about the PyPI source, and this all builds fine.

essandess commented:
> That message is telling you that it is missing `toml` and tries to install that with `pip`, which fails because `pip` isn't there; the solution is to add `py-toml` to the build dependencies instead. Upstream has not listed this explicitly as a dependency, however it's pulled in indirectly through `setuptools-rust`; building in tracemode or on the buildbots without specifying it in this port will make the builds fail. (There are a few other remaining points that I'll review later).

Thanks. There now a few (one) ports that use `setuptools-rust` that have this issue. Pinning this to fix later: https://github.com/macports/macports-ports/blob/master/python/py-tokenizers/Portfile

essandess commented:
> That message is telling you that it is missing `toml` and tries to install that with `pip`, which fails because `pip` isn't there; the solution is to add `py-toml` to the build dependencies instead. Upstream has not listed this explicitly as a dependency, however it's pulled in indirectly through `setuptools-rust`; building in tracemode or on the buildbots without specifying it in this port will make the builds fail. (There are a few other remaining points that I'll review later).

Shouldn't `setuptools-rust` simply add `toml` as a build dependency? Any objections to just adding that fix to this PR?

reneeotten commented:
> > That message is telling you that it is missing `toml` and tries to install that with `pip`, which fails because `pip` isn't there; the solution is to add `py-toml` to the build dependencies instead. Upstream has not listed this explicitly as a dependency, however it's pulled in indirectly through `setuptools-rust`; building in tracemode or on the buildbots without specifying it in this port will make the builds fail. (There are a few other remaining points that I'll review later).
> 
> Shouldn't `setuptools-rust` simply add `toml` as a build dependency? Any objections to just adding that fix to this PR?

It already has, but that doesn't help building in tracemode because `py-toml` isn't specified there or on the buildbots because the ports are being uninstalled in between installations.

reneeotten commented on python/py-cryptography/Portfile:
```diff
@@ -1,49 +1,167 @@
+    depends_lib-append \
+                path:lib/libssl.dylib:openssl \
+                port:py${python.version}-cffi \
+                port:py${python.version}-six
```
this dependency is only needed for version 2.9.2 so it should be moved down into the "legacy" block

essandess replied:
Thanks done.

reneeotten commented on python/py-cryptography/Portfile:
```diff
@@ -1,49 +1,167 @@
+        }
+
+        depends_lib-append \
+                port:py${python.version}-asn1crypto
```
this dependency isn't needed anymore since version 2.8.0 as I pointed out earlier I think

essandess replied:
Thanks done.

reneeotten commented on python/py-cryptography/Portfile:
```diff
@@ -1,49 +1,167 @@
+
+        master_sites \
+                    pypi:[string index ${python.rootname} 0]/${python.rootname}
+        distname    ${python.rootname}-${version}
```
why are you switching now to PyPI for this part and not use GitHub as well?

essandess replied:
Now using git for everything, thanks.

reneeotten commented on python/py-cryptography/Portfile:
```diff
@@ -1,49 +1,167 @@
+                    size    35737283
+
+# https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst
+set cryptography_min_ver 14
```
this isn't very clear for future reference... the point is that the minimum version here should be the OS version where `rust` still builds - correct?

essandess replied:
Clarified thanks.

essandess commented:
> > > That message is telling you that it is missing `toml` and tries to install that with `pip`, which fails because `pip` isn't there; the solution is to add `py-toml` to the build dependencies instead. Upstream has not listed this explicitly as a dependency, however it's pulled in indirectly through `setuptools-rust`; building in tracemode or on the buildbots without specifying it in this port will make the builds fail. (There are a few other remaining points that I'll review later).
> > 
> > 
> > Shouldn't `setuptools-rust` simply add `toml` as a build dependency? Any objections to just adding that fix to this PR?
> 
> It already has, but that doesn't help building in tracemode because `py-toml` isn't specified there or on the buildbots because the ports are being uninstalled in between installations.

I see that @cjones051073 added it as a build dependency. We need it as a run dependency. That means we should add it as a lib dependency, I believe. I will rebase everything to include @cjones051073's commit and add this.

reneeotten commented on python/py-cryptography/Portfile:
add `py-toml` here

essandess replied:
I think toml should be a lib dependency of setuptools-rust, which will fix this issue here and other ports that use it.

reneeotten replied:
if indeed *all* ports that use `setuptools-rust` need also `toml`, then yes that could be a solution. 

It seems a bit similar though to the `setuptools_scm` situation, where sometimes packages specify the dependency as `setuptools_scm[toml]` indicating that `toml` is also needed. The question is then where to add the `toml` dependency; I would say explicitly in the packages that need it, and not in `setuptools_scm` (or `setuptools-rust`), but I can see both sides and don't really care.

essandess replied:
I included a commit to setuptools-rust because I recall this issue in the tokenizers port.

reneeotten reviewed (commented):
hopefully the last round

reneeotten commented on python/py-setuptools-rust/Portfile:
```diff
@@ -31,6 +31,7 @@ if {${name} ne ${subport}} {
     depends_run-append \
                     port:cargo \
                     port:py${python.version}-semantic_version \
+                    port:py${python.version}-toml \
```
this shouldn't be added here since it is not needed at runtime as far as I can tell (and if you would want to add it also as runtime dependency, it should be listed once under `depends_lib` and then remove from `depends_build`)

essandess replied:
Moved to depends_lib

essandess commented:
@reneeotten I believe this is ready to merge. May we all please declare success?

reneeotten commented on python/py-pytest-subtests/Portfile:
```diff
@@ -0,0 +1,43 @@
+description         unittest subTest() support and subtests fixture.
+
+long_description    {*}${description}
+
```
this is missing the `homepage` field as pointed out by `port lint`

reneeotten commented on python/py-pytest-subtests/Portfile:
```diff
@@ -0,0 +1,43 @@
+    # test.run        yes
+    # test.cmd        py.test-${python.branch}
+    # test.target     tests
+    # test.env        PYTHONPATH=${worksrcpath}/build/lib
```
if there are not tests on PyPI (which is fine), just remove this block instead of commenting it out

essandess replied:
Thanks done.

reneeotten commented on python/py-cryptography/Portfile:
this comment is not correct anymore; it says "install pyXY-cryptography for an up-to-date version"; but a user will get this message when exactly doing that when she/he is installing PY27 or on a system that doesn't support `rust`. So saying that it has a CVE in the package is fine, but in the end of the day there isn't anything the user (or we) can do about it.

essandess replied:
Thanks done.

reneeotten commented on python/py-cryptography/Portfile:
```diff
@@ -1,49 +1,161 @@
+                # https://trac.macports.org/ticket/55079
+                patchfiles-append   patch-src__cffi_src_build_openssl.py.diff
+            }
+        }
```
what's the purpose of this block... can this not just be `${os.platform} eq "darwin" && ${os.major} < 11`?

essandess replied:
Legacy code? I changed it to what you suggest.

reneeotten commented on python/py-cryptography/Portfile:
```diff
@@ -1,49 +1,161 @@
+        || ${os.platform} eq "darwin" && ${os.major} < ${cryptography_darwin_min_ver}} {
+        github.setup    pyca cryptography 2.9.2
+
+        description Legacy support of Python cryptography.
```
please align this

essandess replied:
Thanks done.

reneeotten commented on python/py-cryptography/Portfile:
```diff
@@ -1,49 +1,161 @@
+                    size    35737283
+
+# https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst
+set cryptography_darwin_min_ver 14
```
so `rust` still builds on 10.9 according to [this](https://ports.macports.org/port/rust/summary), so please set this to `13` for now and we'll see what the buildbots make of it.

essandess replied:
Thanks done.

reneeotten reviewed (changes requested):
@essandess we can declare almost success I think!
